### ISSUE TITLE
Admin them CSS now has too much specificity

### ISSUE THREAD

**Issue by scleaver** [OrchardCMS/OrchardCore#7830]
Sorry this is long... but it needs a rationale - I like the new dark theme and I don't want to knock all the great work that has been done on the new admin theme but one of the things that has always troubled me about the admin was the high specificity of the way the CSS has been done but with the new admin rewrite it is now so high that I felt it was time to raise the issue.

On upgrade to a release with the new admin recently my own admin theme CSS customization broke not because of changes in the HTML but because of the strength of CSS rules for simple things like links and text.

One of the goals when doing CSS for any project, but particularly for one where you intend to for the users to be able to extend and personalize as needed is to keep the CSS specificity as low as possible across all the selectors in your project. This is so you have plenty of room to override selectors.

OC has been architecture done so well it is just begging to customized and it is easy to create a custom admin theme by using "TheAdmin" as the BaseTheme and throwing in your overrides.

But basically it's now a massive fight and sometimes I'm having to resort to the dreaded `!important` :(

An example, in my admin theme I want to customise the base font and size which should be simple because I am loading my admin css file after TheAdmin theme's css... I should be able to simply go:

```
body {
    font-family: 'Ubuntu', sans-serif;
font-size: 0.875rem;
}
```
But because TheAdmin theme's base font is so specific I can't beat it without getting even more specific which then causes further issues down the line. TheAdmin sets the base font and size like:
```
:root[data-theme="default"] body {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
font-size: 1rem;
}
```
`:root[data-theme="default"]` is totally necessary. I get that there is now a dark theme but the dark theme only needs to modify the color not the font and so the default theme should just be:
```
body {
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
    font-size: 1rem;
color: #212529;
}
```
and the dark theme then should only need to go (without the :root):

```
[data-theme="dark"] body {
color: #7476 ;
}
```

Another example: I want to set the `.ta-navbar-top` to a dark color and the link for `.ta-navbar-brand` it to a light green. This should just be a matter of going:

```
.ta-navbar-brand
{
 color: #32CD32
}
```
But TheAdmin rule is crazy specific so it's not easily overridden.
```
:root[data-theme="default"] .ta-navbar-top .ta-navbar-brand {
    color: rgba(0, 0, 0, 0.9);
}
```
I think you get the idea. If you want more info on the topic have a read of this great article: https://css-tricks.com/strategies-keeping-css-specificity-low/

Personally I think it would have been good to set in place a BEM strategy (http://getbem.com/) or something similar for the backend UI so that developers could easily customise and drastically modify the UI if needed via CSS without having to modify admin templates at much all. But it is probably a bit late in the development for that.

Anyway, I am happy to help on the admin specificity so that the above can be addressed.

**Comment by Skrypt** [comment]
Admin theme has now strict styling due to the fact that it bundles 2 different themes in one.
Moving forward, and based on Bootstrap themselves, this is the way to go to allow having multiple Bootstrap themes color schemes. These changes have been made also, so that the admin theme provides a flicker free experience.

Also, it should not be too complicated to make your custom styles work back.
You could simply try to wrap them around a `:root { }` style which will pass before the `:root[data-theme="something"]` selector in priority. If you want to extend the current Bootstrap theme then your extensions should be included inside the current Admin scss files so that it gets added directly to TheAdmin.css.

If you made custom styles based on Bootstrap directly you can also decide to not have a darkmode theme and use your own theme... or split those 2 themes as 2 css like we had before. But, I can't garantee you that it will be a flicker free experience.
I tried using 2 stylesheets but they are not meant to be switched from javascript ... they tend to execute CSS animations when loading them. You can take a look on Bootstrap Github issues "Support for prefers-color-scheme media query".

**Comment by scleaver** [comment]
Thanks for the reply.

Don't quite follow why having two themes explains why styling has to be so specific thought. You should be able to have many colour themes and still use mostly single selectors for default theme styling like `a { color: #123123; text-decoration: none }` and then `:root[data-theme="orange"] a { color: #123123; text-decoration: none }` only for the overrides... actually not sure why ':root' needs to be used at all as it just adds extra strength to the `[data-theme="orange"]` override that's not necessary. 

Anyway, just thought I will raise it - I can always just not use TheAdmin as the base theme but it is easier just to write a small amount of css on top of the basetheme.

**Comment by Skrypt** [comment]
Yeah but ultimately what we need is to use pure CSS variables to do this and Bootstrap is not there yet. I built something more strict to be able to get there when it happens.

**Comment by sebastienros** [comment]
Is (part of) the issue that you need to prefix all rules with `:root[data-theme="default"]` and `:root[data-theme="dark"]` ?

Also what you suggest, add a default rule, like `a { color: #123123; text-decoration: none }` I assume it's such that you can create a custom Layout that doesn't have the `data-theme` property (@Skrypt mentions there is one) that would then no require `data-theme` selectors. ?

**Comment by Skrypt** [comment]
The issue is that currently we can't detect on server side the current user color-sheme.

Here is one thing I found as a potential future solution.
https://github.com/w3c/csswg-drafts/issues/4162
